File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        def traverse(i , j):
            if i ==0 and j == 0 : return grid[i][j]
            if i== -1 or j == -1 : return -1
            if dp[i][j] != -1 : return dp[i][j]

            left = grid[i][j] + traverse(i , j-1)
            up = grid[i][j]+ traverse(i-1 , j)
            if up == grid[i][j]-1:
                dp[i][j] = left
            elif left == grid[i][j]-1:
                dp[i][j] = up
            else:
                dp[i][j] = min(left , up)

            return dp[i][j]

        dp = [[-1 for i in range(len(grid[0]))] for j in range(len(grid))]

        traverse(len(grid)-1, len(grid[0])-1)
        # print(dp)
        return traverse(len(grid)-1, len(grid[0])-1)

File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:

    int traverse(int i , int j , vector<vector<int>> &grid , vector<vector<int>> &dp){
        if (i ==0 && j == 0 ) return grid[i][j];

        int up = INT_MAX , left = INT_MAX;
        if (dp[i][j] != -1) return dp[i][j];
        if (i-1 >=0) up = grid[i][j] + traverse(i-1 , j , grid , dp);
        if (j-1 >=0)left = grid[i][j] + traverse(i , j-1 , grid , dp); 
        
        return dp[i][j] = min(up , left );

    }
    int minPathSum(vector<vector<int>>& grid) {
        vector<vector<int>> dp(grid.size() +1,vector<int>(grid[0].size(),-1));
        return traverse(grid.size()-1 , grid[0].size()-1 , grid , dp);
    }
};